src/dataflow/src/server.rs

```diff
@@ -218,19 +217,216 @@ pub fn serve(config: Config) -> Result<WorkerGuards<()>, String> {
                 pending_peeks: Vec::new(),
                 feedback_tx: None,
                 reported_frontiers: HashMap::new(),
+                reported_source_frontiers: HashMap::new(),
                 metrics: Metrics::for_worker_id(worker_idx),
             }
             .run()
         },
     )
 }
 
+/// This struct holds per-source timestamp state in a way that can be shared across
+/// different source instances and allow different source instances to indicate
+/// how far they have read up to.
+///
+/// This type is almost never meant to be used directly, and you probably want to
+/// use `TimestampBindingRc` instead.
+pub struct TimestampBindingBox {
+    /// List of timestamp bindings per independent partition. This vector is sorted
+    /// by timestamp and offset and each `(time, offset)` entry indicates that offsets <=
+    /// `offset` should be assigned `time` as their timestamp. Consecutive entries form
+    /// an interval of offsets.
+    partitions: HashMap<PartitionId, Vec<(Timestamp, MzOffset)>>,
+    /// Indicates the lowest timestamp across all partitions that we retain bindings for.
+    /// This frontier can be held back by other entities holding the shared
+    /// `TimestampBindingRc`.
+    compaction_frontier: MutableAntichain<Timestamp>,
+}
+
+impl TimestampBindingBox {
+    /// Create a new instance of `TimestampBindingBox`.
+    fn new() -> Self {
+        Self {
+            partitions: HashMap::new(),
+            compaction_frontier: MutableAntichain::new(),
+        }
+    }
+
+    fn adjust_compaction_frontier(
+        &mut self,
+        remove: AntichainRef<Timestamp>,
+        add: AntichainRef<Timestamp>,
+    ) {
+        self.compaction_frontier
+            .update_iter(remove.iter().map(|t| (*t, -1)));
+        self.compaction_frontier
+            .update_iter(add.iter().map(|t| (*t, 1)));
+    }
+
+    /// Drop all of the timestamp bindings for timestamps not in advance of the
+    /// compaction frontier.
+    fn compact(&mut self) {
+        let frontier = self.compaction_frontier.frontier();
+
+        if !frontier.is_empty() {
+            // Other folks are using these timestamp bindings. Lets keep everything
+            // >= what people are using
+            for (_, entries) in self.partitions.iter_mut() {
+                entries.retain(|(time, _)| frontier.less_equal(time));
+            }
+        }
+    }
```

Comment:
I'm not clear on the behavior here. It seems like the intent is to produce a compacted representation of the intervals represented in the struct, but this probably doesn't do that if e.g. the last statement is not at a time that is beyond the frontier. That seems like a totally plausible thing to happen if one part stops emitting results for a while.

I can see a few fixes, but I'd recommend one that advances each of the times in `entries` and then fuses intervals that are compatible (same time, keep the latter offset). The difference here would be that the last offset would have its time advanced, but would not be discarded. It also seems to generalize nicely if you even want the intervals to be explicit, so that you could have non-contiguous offsets for a time.

Reply:
Or, sorry without prescribing a solution, I'd first jot down the intended property (e.g. "the source can be replayed in the context of any `as_of` that is beyond `self.compaction_frontier`").

Reply:
done, added better documentation and compacted in the manner you describe. thanks for the catch!